File: pushbot-go/device.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"
)

type DeviceInfo struct {
	IP        string
	MAC       string
	Name      string
	Timestamp int64
	Interface string
}

func readIPAddress(path string) []DeviceInfo {
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	var list []DeviceInfo
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		parts := strings.Fields(line)
		if len(parts) >= 4 {
			ts, _ := strconv.ParseInt(parts[3], 10, 64)
			inf := ""
			if len(parts) >= 5 {
				inf = parts[4]
			}
			list = append(list, DeviceInfo{
				IP: parts[0], MAC: parts[1], Name: parts[2],
				Timestamp: ts, Interface: inf,
			})
		}
	}
	return list
}

func writeIPAddress(path string, list []DeviceInfo) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	for _, d := range list {
		_, _ = fmt.Fprintf(f, "%s %s %s %d %s\n", d.IP, d.MAC, d.Name, d.Timestamp, d.Interface)
	}
	return nil
}

// getMAC 从 ipAddress、tmp_downlist、dhcp.leases、arp 获取 MAC
func getMAC(cfg *Config, ip string) string {
	dir := cfg.Dir
	// 缓存
	for _, p := range []string{filepath.Join(dir, "ipAddress"), filepath.Join(dir, "tmp_downlist")} {
		if mac := grepField(p, ip, 2); mac != "" {
			return mac
		}
	}
	if mac := grepField("/var/dhcp.leases", ip, 2); mac != "" {
		return mac
	}
	if mac := arpGetMAC(ip); mac != "" {
		return mac
	}
	return "unknown"
}

func grepField(path, matchCol1 string, fieldIndex int) string {
	f, err := os.Open(path)
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) > fieldIndex && parts[0] == matchCol1 {
			return parts[fieldIndex-1]
		}
	}
	return ""
}

func arpGetMAC(ip string) string {
	f, err := os.Open("/proc/net/arp")
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Scan() // header
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) >= 6 && parts[0] == ip && (parts[2] == "0x2" || parts[2] == "0x6") {
			mac := parts[3]
			if mac != "00:00:00:00:00:00" {
				return mac
			}
		}
	}
	return ""
}

// getName 从 device_aliases、ipAddress、dhcp、UCI dhcp、OUI 获取名称
func getName(cfg *Config, ip, mac string) string {
	for _, line := range cfg.DeviceAliases {
		idx := strings.Index(line, " ")
		if idx <= 0 {
			continue
		}
		m := strings.TrimSpace(line[:idx])
		if strings.EqualFold(m, mac) {
			return strings.TrimSpace(line[idx+1:])
		}
	}
	dir := cfg.Dir
	for _, p := range []string{filepath.Join(dir, "ipAddress"), filepath.Join(dir, "tmp_downlist")} {
		if n := grepField(p, ip, 3); n != "" {
			return n
		}
	}
	if n := grepField("/var/dhcp.leases", ip, 4); n != "" {
		return n
	}
	// UCI dhcp 可选
	if n := uciDHCPName(ip); n != "" {
		return n
	}
	if mac != "unknown" {
		if n := ouiLookup(mac, cfg); n != "" {
			return n
		}
	}
	return "unknown"
}

func uciDHCPName(ip string) string {
	cmd := exec.Command("uci", "show", "dhcp")
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	lines := strings.Split(string(out), "\n")
	for _, l := range lines {
		if !strings.Contains(l, ".ip=") && !strings.Contains(l, ".ip='") {
			continue
		}
		if !strings.Contains(l, ip) {
			continue
		}
		i := strings.Index(l, "=")
		if i <= 0 {
			continue
		}
		key := strings.TrimSpace(l[:i])
		// key 形如 dhcp.@host[0].ip
		sec := strings.TrimSuffix(key, ".ip")
		if sec == key {
			continue
		}
		name := uciGetFull(sec + ".name")
		if name != "" {
			return name
		}
	}
	return ""
}

func uciGetFull(path string) string {
	cmd := exec.Command("uci", "-q", "get", path)
	out, _ := cmd.Output()
	return strings.TrimSpace(string(out))
}

// ouiSearchPaths 查找顺序：/usr/share/pushbot（包安装）→ 配置目录 → 临时目录
var ouiSearchPaths = []string{"/usr/share/pushbot", ""}

func ouiLookup(mac string, cfg *Config) string {
	if len(mac) < 8 {
		return ""
	}
	oui := strings.ToLower(strings.ReplaceAll(mac[:8], ":", ""))
	bases := make([]string, 0, 4)
	for _, p := range ouiSearchPaths {
		if p != "" {
			bases = append(bases, p)
		}
	}
	bases = append(bases, cfg.ConfigDir, cfg.Dir)
	for _, base := range bases {
		for _, name := range []string{"oui_base.txt", "oui.txt"} {
			ouiPath := filepath.Join(base, name)
			f, err := os.Open(ouiPath)
			if err != nil {
				continue
			}
			sc := bufio.NewScanner(f)
			for sc.Scan() {
				line := sc.Text()
				idx := strings.Index(line, "(base 16)")
				if idx <= 0 {
					continue
				}
				// 行首 OUI 可能为 XX-XX-XX 或 XX:XX:XX，统一与 oui 比较
				pre := line
				if idx < len(line) {
					pre = line[:idx]
				}
				prefix := strings.ReplaceAll(strings.ReplaceAll(pre, "-", ""), ":", "")
				prefix = strings.ToLower(strings.TrimSpace(prefix))
				if len(prefix) >= 6 && prefix[:6] == oui {
					rest := line[idx+9:]
					parts := strings.Split(rest, "\t")
					name := strings.TrimSpace(parts[len(parts)-1])
					f.Close()
					return strings.ReplaceAll(name, " ", "_")
				}
			}
			f.Close()
		}
	}
	return ""
}

// getInterface 从 ipAddress、arp、iw 获取接口
func getInterface(cfg *Config, mac string) string {
	dir := cfg.Dir
	if s := grepFieldByMAC(filepath.Join(dir, "ipAddress"), mac, 5); s != "" {
		return s
	}
	if s := grepFieldByMAC(filepath.Join(dir, "tmp_downlist"), mac, 5); s != "" {
		return s
	}
	f, err := os.Open("/proc/net/arp")
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Scan()
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) >= 6 && strings.EqualFold(parts[3], mac) {
			return parts[5]
		}
	}
	return ""
}

func grepFieldByMAC(path, mac string, fieldIndex int) string {
	f, err := os.Open(path)
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) >= fieldIndex && strings.EqualFold(parts[1], mac) {
			return parts[fieldIndex-1]
		}
	}
	return ""
}

// blackwhitelist 返回 0 表示应推送（通过检查），非 0 表示不推送
func blackwhitelist(cfg *Config, mac string) int {
	if cfg.PushbotWhitelist != "" || cfg.PushbotBlacklist != "" || cfg.PushbotInterface != "" ||
		cfg.MACOnlineList != "" || cfg.MACOfflineList != "" {
		// 白名单：仅在名单内才推送
		for _, m := range cfg.PushbotWhitelistLines {
			if strings.TrimSpace(m) != "" && strings.EqualFold(m, mac) {
				return 0
			}
		}
		if cfg.PushbotWhitelist != "" {
			return 1
		}
		// 黑名单：在名单内不推送
		for _, m := range cfg.PushbotBlacklistLines {
			if strings.TrimSpace(m) != "" && strings.EqualFold(m, mac) {
				return 1
			}
		}
		// 接口过滤等可再扩展
	}
	return 0
}

// pingOK 检测 IP 是否在线（arping 或 ping）
func pingOK(ip string, timeoutSec, retry int) bool {
	// 先尝试 arping
	if iface := arpGetIface(ip); iface != "" {
		cmd := exec.Command("arping", "-I", iface, "-c", "3", "-w", strconv.Itoa(timeoutSec), ip)
		cmd.Run()
		if cmd.ProcessState != nil && cmd.ProcessState.ExitCode() == 0 {
			return true
		}
	}
	for i := 0; i < retry; i++ {
		cmd := exec.Command("ping", "-c", "2", "-W", strconv.Itoa(timeoutSec), ip)
		if cmd.Run() == nil {
			return true
		}
		time.Sleep(time.Second)
	}
	return false
}

func arpGetIface(ip string) string {
	f, _ := os.Open("/proc/net/arp")
	if f == nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Scan()
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) >= 6 && parts[0] == ip {
			return parts[5]
		}
	}
	return ""
}

// 扫描当前在线设备并更新 ipAddress；发送上线通知
func pushbotFirst(app *App) {
	cfg := app.cfg
	dir := cfg.Dir
	ipPath := filepath.Join(dir, "ipAddress")
	current := readIPAddress(ipPath)

	// 一次读取 ARP，得到候选 IP 与 ip->mac，避免每 IP 再读一次
	arpIPs, ipToMac := readArpOnce()
	var stillOnline []DeviceInfo
	for _, d := range current {
		if pingOK(d.IP, cfg.DownTimeout, cfg.TimeoutRetry) {
			stillOnline = append(stillOnline, d)
		} else {
			app.tmpDownMu.Lock()
			app.tmpDownList = append(app.tmpDownList, d)
			app.tmpDownMu.Unlock()
		}
	}
	var newList []DeviceInfo
	seenIP := make(map[string]bool)
	for _, d := range stillOnline {
		seenIP[d.IP] = true
		newList = append(newList, d)
	}
	for _, ip := range arpIPs {
		if seenIP[ip] {
			continue
		}
		if !pingOK(ip, cfg.UpTimeout, cfg.TimeoutRetry) {
			continue
		}
		mac := ipToMac[ip]
		if mac == "" {
			mac = getMAC(cfg, ip)
		}
		name := getName(cfg, ip, mac)
		iface := getInterface(cfg, mac)
		if blackwhitelist(cfg, mac) != 0 {
			continue
		}
		seenIP[ip] = true
		newList = append(newList, DeviceInfo{
			IP: ip, MAC: mac, Name: name,
			Timestamp: time.Now().Unix(), Interface: iface,
		})
		if cfg.PushbotUp == 1 {
			app.appendNotify(notifyUp(ip, mac, name, iface, cfg))
		}
	}
	_ = writeIPAddress(ipPath, newList)
}

// readArpOnce 一次读取 /proc/net/arp，返回候选 IP 列表和 ip->mac 映射，避免多次打开文件
func readArpOnce() (ips []string, ipToMac map[string]string) {
	f, err := os.Open("/proc/net/arp")
	if err != nil {
		return nil, nil
	}
	defer f.Close()
	ipToMac = make(map[string]string)
	sc := bufio.NewScanner(f)
	sc.Scan()
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) < 6 || (parts[2] != "0x2" && parts[2] != "0x6") {
			continue
		}
		ip, mac := parts[0], parts[3]
		if strings.HasPrefix(ip, "169.254.") || mac == "00:00:00:00:00:00" {
			continue
		}
		ips = append(ips, ip)
		ipToMac[ip] = mac
	}
	return ips, ipToMac
}

func arpIPList() []string {
	ips, _ := readArpOnce()
	return ips
}

// 处理离线列表并追加下线通知
func downSend(app *App) {
	app.tmpDownMu.Lock()
	list := app.tmpDownList
	app.tmpDownList = nil
	app.tmpDownMu.Unlock()
	cfg := app.cfg
	for _, d := range list {
		if blackwhitelist(cfg, d.MAC) != 0 {
			continue
		}
		if cfg.PushbotDown != 1 {
			continue
		}
		onlineSec := time.Now().Unix() - d.Timestamp
		app.appendNotify(notifyDown(d.IP, d.MAC, d.Name, timeForHumans(int(onlineSec)), cfg))
	}
}

var appNotifyMu sync.Mutex

func (a *App) appendNotify(n *Notify) {
	if n == nil {
		return
	}
	appNotifyMu.Lock()
	if a.pendingTitle == "" {
		a.pendingTitle = n.Title
	}
	a.pendingContent += n.Content
	appNotifyMu.Unlock()
	// 有待发通知时唤醒主循环，避免空转轮询
	if a.wakeCh != nil {
		select {
		case a.wakeCh <- struct{}{}:
		default:
		}
	}
}

func (a *App) consumeNotify() (title, content string) {
	appNotifyMu.Lock()
	title, content = a.pendingTitle, a.pendingContent
	a.pendingTitle = ""
	a.pendingContent = ""
	appNotifyMu.Unlock()
	return title, content
}

File: pushbot-c/device.c
#include "common.h"
#include "config.h"
#include "device.h"
#include "util.h"
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static int ping_ok(const char *ip, int timeout_sec, int retry, const char *iface) {
	char cmd[384];
	int i;
	if (iface && iface[0]) {
		snprintf(cmd, sizeof(cmd), "arping -I %s -c 3 -w %d %s 2>/dev/null", iface, timeout_sec, ip);
		if (system(cmd) == 0) return 1;
	}
	for (i = 0; i < retry; i++) {
		snprintf(cmd, sizeof(cmd), "ping -c 2 -W %d %s 2>/dev/null", timeout_sec, ip);
		if (system(cmd) == 0) return 1;
		sleep(1);
	}
	return 0;
}

static void arp_get_iface(const char *ip, char *iface_out, size_t sz) {
	FILE *f = fopen("/proc/net/arp", "r");
	if (!f) { if (iface_out && sz) iface_out[0] = '\0'; return; }
	char line[256];
	if (!fgets(line, sizeof(line), f)) { fclose(f); return; }
	while (fgets(line, sizeof(line), f)) {
		char tip[64];
		char iface[32];
		if (sscanf(line, "%63s %*s %*s %*s %*s %31s", tip, iface) >= 2 && strcmp(tip, ip) == 0) {
			strncpy(iface_out, iface, sz - 1);
			iface_out[sz - 1] = '\0';
			fclose(f);
			return;
		}
	}
	fclose(f);
	if (iface_out && sz) iface_out[0] = '\0';
}

static void get_mac_from_arp(const char *ip, char *mac_out, size_t sz) {
	FILE *f = fopen("/proc/net/arp", "r");
	if (!f) { if (mac_out && sz) mac_out[0] = '\0'; return; }
	char line[256];
	fgets(line, sizeof(line), f);
	while (fgets(line, sizeof(line), f)) {
		char tip[64], tmac[32];
		if (sscanf(line, "%63s %*s %*s %31s", tip, tmac) >= 2 && strcmp(tip, ip) == 0) {
			if (strcmp(tmac, "00:00:00:00:00:00") != 0) {
				strncpy(mac_out, tmac, sz - 1);
				mac_out[sz - 1] = '\0';
			}
			break;
		}
	}
	fclose(f);
}

static void get_name_from_dhcp(const char *ip, char *name_out, size_t sz) {
	char cmd[320];
	snprintf(cmd, sizeof(cmd), "grep -F '%s' /tmp/dhcp.leases 2>/dev/null | awk '{print $4}'", ip);
	FILE *f = popen(cmd, "r");
	if (!f) { if (name_out && sz) name_out[0] = '\0'; return; }
	if (fgets(name_out, (int)sz, f)) {
		size_t len = strlen(name_out);
		if (len > 0 && name_out[len-1] == '\n') name_out[len-1] = '\0';
	} else
		name_out[0] = '\0';
	pclose(f);
}

static void oui_lookup(const char *mac, const config_t *cfg, char *name_out, size_t sz) {
	if (strlen(mac) < 8) { if (name_out && sz) name_out[0] = '\0'; return; }
	char oui[16];
	snprintf(oui, sizeof(oui), "%.2s%.2s%.2s", mac, mac+3, mac+6);
	for (int i = 0; oui[i]; i++) oui[i] = (char)tolower((unsigned char)oui[i]);
	const char *dirs[] = { "/usr/share/pushbot", cfg->config_dir, cfg->dir, NULL };
	for (int d = 0; dirs[d]; d++) {
		char path[256];
		snprintf(path, sizeof(path), "%s/oui_base.txt", dirs[d]);
		FILE *f = fopen(path, "r");
		if (!f) { snprintf(path, sizeof(path), "%s/oui.txt", dirs[d]); f = fopen(path, "r"); }
		if (!f) continue;
		char line[512];
		while (fgets(line, sizeof(line), f)) {
			if (!strstr(line, "(base 16)")) continue;
			char *p = strstr(line, oui);
			if (!p || p > line + 16) continue;
			p = strstr(line, "(base 16)");
			if (!p) continue;
			p += 9;
			while (*p == '\t' || *p == ' ') p++;
			char *end = strchr(p, '\n');
			if (end) *end = '\0';
			/* copy and replace space with _ */
			size_t i = 0;
			for (; *p && i < sz - 1; p++) {
				name_out[i++] = (*p == ' ') ? '_' : *p;
			}
			name_out[i] = '\0';
			fclose(f);
			return;
		}
		fclose(f);
	}
	if (name_out && sz) name_out[0] = '\0';
}

void get_name(const config_t *cfg, const char *ip, const char *mac, char *name_out, size_t sz) {
	get_name_from_dhcp(ip, name_out, sz);
	if (name_out[0]) return;
	oui_lookup(mac, cfg, name_out, sz);
	if (name_out[0]) return;
	snprintf(name_out, sz, "%s", mac);
}

static int in_whitelist(const config_t *cfg, const char *mac) {
	if (!cfg->pushbot_whitelist[0]) return 0;
	char list[MAX_LINE];
	strncpy(list, cfg->pushbot_whitelist, sizeof(list)-1);
	list[sizeof(list)-1] = '\0';
	for (char *p = strtok(list, " \n\t"); p; p = strtok(NULL, " \n\t"))
		if (strcasecmp(p, mac) == 0) return 1;
	return 0;
}

static int in_blacklist(const config_t *cfg, const char *mac) {
	if (!cfg->pushbot_blacklist[0]) return 0;
	char list[MAX_LINE];
	strncpy(list, cfg->pushbot_blacklist, sizeof(list)-1);
	list[sizeof(list)-1] = '\0';
	for (char *p = strtok(list, " \n\t"); p; p = strtok(NULL, " \n\t"))
		if (strcasecmp(p, mac) == 0) return 1;
	return 0;
}

int blackwhitelist(const config_t *cfg, const char *mac) {
	if (cfg->pushbot_whitelist[0]) return in_whitelist(cfg, mac) ? 0 : 1;
	if (in_blacklist(cfg, mac)) return 1;
	return 0;
}

int device_read_list(const char *path, device_info_t *list, int max_n, int *n) {
	*n = 0;
	FILE *f = fopen(path, "r");
	if (!f) return 0;
	char line[512];
	while (fgets(line, sizeof(line), f) && *n < max_n) {
		char ip[64], mac[32], name[128], ts[32], iface[32];
		if (sscanf(line, "%63s %31s %127[^\t\n] %31s %31s", ip, mac, name, ts, iface) >= 2) {
			device_info_t *d = &list[*n];
			strncpy(d->ip, ip, sizeof(d->ip)-1);
			strncpy(d->mac, mac, sizeof(d->mac)-1);
			strncpy(d->name, name, sizeof(d->name)-1);
			strncpy(d->iface, iface, sizeof(d->iface)-1);
			d->timestamp = (time_t)atoll(ts);
			(*n)++;
		}
	}
	fclose(f);
	return 1;
}

void device_write_list(const char *path, const device_info_t *list, int n) {
	FILE *f = fopen(path, "w");
	if (!f) return;
	for (int i = 0; i < n; i++) {
		fprintf(f, "%s\t%s\t%s\t%ld\t%s\n",
			list[i].ip, list[i].mac, list[i].name,
			(long)list[i].timestamp, list[i].iface);
	}
	fclose(f);
}

typedef struct arp_entry { char ip[64]; char mac[32]; } arp_entry_t;
static int read_arp_once(arp_entry_t *out, int max_n) {
	FILE *f = fopen("/proc/net/arp", "r");
	if (!f) return 0;
	char line[256];
	int n = 0;
	fgets(line, sizeof(line), f);
	while (fgets(line, sizeof(line), f) && n < max_n) {
		char ip[64], mac[32], flags[8];
		if (sscanf(line, "%63s %*s %*s %31s %*s %*s", ip, mac, flags) < 2) continue;
		if (strncmp(ip, "169.254.", 8) == 0) continue;
		if (strcmp(mac, "00:00:00:00:00:00") == 0) continue;
		if (strcmp(flags, "0x2") != 0 && strcmp(flags, "0x6") != 0) continue;
		strncpy(out[n].ip, ip, sizeof(out[n].ip)-1);
		strncpy(out[n].mac, mac, sizeof(out[n].mac)-1);
		n++;
	}
	fclose(f);
	return n;
}

static void get_interface_for_mac(const config_t *cfg, const char *mac, char *iface_out, size_t sz) {
	(void)cfg;
	FILE *f = fopen("/proc/net/arp", "r");
	if (!f) { if (iface_out && sz) iface_out[0] = '\0'; return; }
	char line[256];
	fgets(line, sizeof(line), f);
	while (fgets(line, sizeof(line), f)) {
		char tmac[32], iface[32];
		if (sscanf(line, "%*s %*s %*s %31s %*s %31s", tmac, iface) >= 2 && strcasecmp(tmac, mac) == 0) {
			strncpy(iface_out, iface, sz - 1);
			iface_out[sz - 1] = '\0';
			fclose(f);
			return;
		}
	}
	fclose(f);
	if (iface_out && sz) iface_out[0] = '\0';
}

void pushbot_first(config_t *cfg, const char *ip_path, const char *log_path,
	device_info_t *current, int n_current,
	device_info_t *new_list, int *n_new,
	notify_cb on_up, notify_cb on_down, void *cb_ctx) {
	arp_entry_t arp[128];
	int n_arp = read_arp_once(arp, 128);
	device_info_t still[MAX_DEVICES];
	int n_still = 0;
	int seen[128];
	memset(seen, 0, sizeof(seen));

	for (int i = 0; i < n_current; i++) {
		char iface[32];
		arp_get_iface(current[i].ip, iface, sizeof(iface));
		if (ping_ok(current[i].ip, cfg->down_timeout, cfg->timeout_retry, iface)) {
			still[n_still++] = current[i];
		} else {
			if (on_down) on_down(cb_ctx, &current[i], NULL);
		}
	}

	for (int i = 0; i < n_still; i++)
		new_list[*n_new++] = still[i];

	for (int a = 0; a < n_arp && *n_new < MAX_DEVICES; a++) {
		int j;
		for (j = 0; j < *n_new; j++)
			if (strcmp(new_list[j].ip, arp[a].ip) == 0) break;
		if (j < *n_new) continue;
		char iface[32];
		arp_get_iface(arp[a].ip, iface, sizeof(iface));
		if (!ping_ok(arp[a].ip, cfg->up_timeout, cfg->timeout_retry, iface)) continue;
		if (blackwhitelist(cfg, arp[a].mac)) continue;
		device_info_t d;
		strncpy(d.ip, arp[a].ip, sizeof(d.ip)-1);
		strncpy(d.mac, arp[a].mac, sizeof(d.mac)-1);
		get_name(cfg, d.ip, d.mac, d.name, sizeof(d.name));
		get_interface_for_mac(cfg, d.mac, d.iface, sizeof(d.iface));
		d.timestamp = time(NULL);
		new_list[*n_new] = d;
		(*n_new)++;
		if (cfg->pushbot_up && on_up) on_up(cb_ctx, &d, NULL);
	}
}

int device_ping_ok(const char *ip, int timeout_sec, int retry) {
	char iface[32];
	arp_get_iface(ip, iface, sizeof(iface));
	return ping_ok(ip, timeout_sec, retry, iface);
}
